neural_compressor/adaptor/torch_utils/gptq.py

```diff
@@ -0,0 +1,586 @@
+import math
+import time
+
+import torch
+import torch.nn as nn
+import transformers
+from tqdm import tqdm
+from functools import partial
+# different models may have different input structures. 
+# arch_inputs = {
+#     'opt': ['attention_mask'],
+#     'bloom': ['attention_mask', 'alibi'],
+#     'llama': ['attention_mask', 'position_ids']
+# }
+
+DEBUG = False 
+
+def is_leaf(module):
+    children_cnt = 0
+    for n in module.children():
+        children_cnt += 1
+    return True if children_cnt == 0 else False
+
+def trace_gptq_target_blocks(module, module_types = [torch.nn.ModuleList]):
+    """
+    seperately trace two parts related to gptq:
+    1. embedding layers (layers before transformer stacks)
+    2. transformer stacks
+    """
+    gptq_related_blocks = {
+        "embeddings": {},
+        "transformers_pre": {}, # todo
+        "transformers_name": None,
+        "transformers": None,
+        "transformers": {}, # todo
+    }
+    for n, m in module.named_modules():
+        if type(m) in module_types:
+            gptq_related_blocks["transformers_name"] = n
+            gptq_related_blocks["transformers"] = m
+            return gptq_related_blocks
+        else:
+            if is_leaf(m):
+                gptq_related_blocks["embeddings"][n] = m
+    return gptq_related_blocks
+
+def find_layers(module, layers=[nn.Conv2d, nn.Linear], name=''):
+    if type(module) in layers:
+        return {name: module}
+    res = {}
+    for name1, child in module.named_children():
+        res.update(find_layers(
+            child, layers=layers, name=name + '.' + name1 if name != '' else name1
+        ))
+    return res
+
+def quantize(x, scale, zero, maxq):
+    if maxq < 0:
+        return (x > scale / 2).float() * scale + (x < zero / 2).float() * zero
+    q = torch.clamp(torch.round(x / scale) + zero, 0, maxq)
+    return scale * (q - zero)
+    
+class InputHooker(nn.Module):
+    def __init__(self, module, arch, inp, cache):
+        super().__init__()
+        self.module = module
+        self.inp = inp
+        self.cache = cache
+        # self.input_args = arch_inputs[arch]
+    def forward(self, inp, **kwargs):
+        # import pdb;pdb.set_trace()
+        self.inp[self.cache['i']] = inp
+        self.cache['i'] += 1
+        for arg in kwargs:
+            if isinstance(kwargs[arg], torch.Tensor):
+                self.cache[arg] = kwargs[arg]
+            else:
+                continue
+        print("input hooked successfully")
+        raise ValueError
+
+class GPTQuantizer(object):
+    """Main API"""
+    def __init__(self, model, dataloader, device, args):

```

Comment:
Since lots of the codes are inspired by GPTQ, please check with suyue for  the license issue, and at least add docstring to appreciate their efforts